### PR TITLE
Propose Solana ABI management

### LINKED ISSUE #7421: Design Solana ABI management process
The Solana ABI (binary interface to the cluster) is currently only defined implicitly by the implementation, and requires a very careful eye to notice breaking changes.   This makes it extremely difficult to upgrade the software on an existing cluster without rebooting the ledger.  During early development we deliberately decided to ignore breaking ABI changes but we can no longer afford this luxury.

As a starting point, I think we will need: 
1) A way to identify all structs/enums that cannot be modified without breaking the ABI (Transactions, Shreds, Ledger bits).
2) A mechanism to turn CI red when those structs/enums are modified. 
3) A human process to decide whether a breaking change should be accepted, and if so how to manage the resulting hard fork.
5) A comprehensive test suite (cc: #6123) of golden binary ledgers that should always be consumable by newer software versions
4) Guidelines to the eng team for all of the above, including general agreement on how strict we want to be.  eg, once the system program is published/locked down, would it be permitted to even add a new system instruction?   Most strictly, no

### PR DESCRIPTION
**Pull request by ryoqun**
#### Problem

Solana's ABI management is non-existent, yet we will definitely need one [according to this](https://github.com/solana-labs/solana/issues/7421#issue-536162047):

> The Solana ABI (binary interface to the cluster) is currently only defined implicitly by the implementation, and requires a very careful eye to notice breaking changes. This makes it extremely difficult to upgrade the software on an existing cluster without rebooting the ledger. During early development we deliberately decided to ignore breaking ABI changes but we can no longer afford this luxury.

#### Summary of Changes

Added a (draft) proposal for it.

Fixes #7421

### CODE REVIEW

**Comment by ryoqun** [comment]
@mvines This is still half done, but I want to gather feedback as early as possible. This is what I'm currently have in my mind. Let alone flacky markdown formatting, no grammarly pass, this currently even omits to cover some points mentioned in the original issue. But I want to confirm general direction of this proposal is good or needs adjustment.

File: book/src/proposals/abi-management.md [hunk 1]
```diff
@@ -0,0 +1,434 @@
...
+
+run `cargo test` with specific environment like `SOLANA_UPDATE_ABI_FIXTURES=1`.
+Then, these fixtures will be automatically updated and review with `git diff`
+
+# Implemention remarks
+
+We'll end up heavily relying on macro magics. For precedent, `ink' from the
+Parity Technologies [2], could be informational.
+
+# Considerations
```
Comment (ryoqun) [inline book/src/proposals/abi-management.md]: I want these remaining concern to be clarified or I should propose a concrete proposal for each. :)

File: book/src/proposals/abi-management.md [hunk 2]
```diff
@@ -0,0 +1,434 @@
...
+
+The attribute try to detect any unsanctioned changes to the marked ABI-related
+things when tests are run locally or on the CI, gathering closer attentions
+from engineers further on. However, the detection cannot be complete; no matter
+how hard we statically analyze the source code, it's still possible to break
+ABI.
+
+Thus, we confirm non-breakage by running paired tests for each ABI-related
+things which are separated for both serialization and deserialization and are
+versioned by the use of test fixtures containing actual binaries.
```
Comment (ryoqun) [inline book/src/proposals/abi-management.md]: The overall framing/scope of this design proposal is legit/sound? I think this is what we wanted from [the original comment](https://github.com/solana-labs/solana/pull/7415#discussion_r356392307), is this right?

File: book/src/proposals/abi-management.md [hunk 3]
```diff
@@ -0,0 +1,434 @@
...
+     pub hash: Hash,
++    #[solana_abi_added(since = "rev1")]
++    /// processing timestamp of last slot
++    pub timestamp: Option<UnixTimestamp>,
+ }
+
+ impl SolanaABI for Vote {
+   const REVISION_LIST = [
+      (0, "epoch", 0..), // leave this revision acceptable for indefinite time
++     (1, "epoch", 21..),
```
Comment (ryoqun) [inline book/src/proposals/abi-management.md]: Is this level of ABI management is too much grained? I think this can be justified by the finer and dense unit testing. But it could be regarded as an overkill. This is my preference, and there are seems no projects doing similar practice.

File: book/src/proposals/abi-management.md [hunk 4]
```diff
@@ -0,0 +1,434 @@
...
+The Solana ABI (binary interface to the cluster) is currently only defined 
+implicitly by the implementation, and requires a very careful eye to notice
+breaking changes. This makes it extremely difficult to upgrade the software
+on an existing cluster without rebooting the ledger.
+
+# Requirements and objectives
+
+- Unintended ABI changes can be detected as CI failures mechanically.
+- Newer impl must be able to process oldest data (since genesis) once we go
+  mainnet. (This is in stark contrast with regard to conventional software;
```
Comment (ryoqun) [inline book/src/proposals/abi-management.md]: I hope this holds true for DLT... But might be wrong?

File: book/src/proposals/abi-management.md [hunk 5]
```diff
@@ -0,0 +1,434 @@
...
+
+# Implemention remarks
+
+We'll end up heavily relying on macro magics. For precedent, `ink' from the
+Parity Technologies [2], could be informational.
+
+# Considerations
+ 
+- how to detect vesions?
+  - Introduce version header? (2 byte for everything)
```
Comment (ryoqun) [inline book/src/proposals/abi-management.md]: This is tough question for me as a rather newbie here, by the way... ;)

File: book/src/proposals/abi-management.md [hunk 6]
```diff
@@ -0,0 +1,434 @@
...
+
+# Hard fork work flow
+
+(To be documented.)
+
+# Complehensive test suite of golden binary ledgers
+
+(To be documented: Maintain ledger generation script which executes tx once for
+each program?  While we should verify we can consume older ledgers, we also
+should verify we can produce logically identical ledgers from scratch.)
```
Comment (ryoqun) [inline book/src/proposals/abi-management.md]: If this is the preferred approach to explore the ABI problem, I'll quickly shift my work weight. :)

File: book/src/proposals/abi-management.md [hunk 7]
```diff
@@ -0,0 +1,434 @@
...
+```
+
+# Compatibility strictness
+
+(To be documented: table of compatibility of compatibility item types, status
+and kind of changes)
+
+# Hard fork work flow
+
+(To be documented.)
```
Comment (ryoqun) [inline book/src/proposals/abi-management.md]: Because this is draft, I deliberately left this vacant. As a next step, should I fill and propose this part as soon as possible?

File: book/src/proposals/abi-management.md [hunk 8]
```diff
@@ -0,0 +1,434 @@
...
+   // no change for ABI test is needed.
+Files:
+ .../fixtures/stored-account-rev-0.hex.yaml
++.../fixtures/stored-account-rev-1.hex.yaml
+```
+
+# Compatibility strictness
+
+(To be documented: table of compatibility of compatibility item types, status
+and kind of changes)
```
Comment (ryoqun) [inline book/src/proposals/abi-management.md]: Also, as for `stirctnness`, I have yet to answer to this original point from [the issue](#7421)...:

> general agreement on how strict we want to be. eg, once the system program is published/locked down, would it be permitted to even add a new system instruction? Most strictly, no

File: book/src/proposals/abi-management.md [hunk 9]
```diff
@@ -0,0 +1,434 @@
...
+
+As far as I researched, I couldn't find good proven standard approach to the ABI
+management, to my surprise. It seems that ad-hoc or integration-test level ABI
+management is common in the scene. However, I think these systematic management
+will provide confidene of changes and allow rapid development.
+
+# References
+
+1. Bitcoin Wiki: https://en.bitcoin.it/wiki/Hardfork
+2. Prity's ink to write smart contracts: https://github.com/paritytech/ink
```
Comment (garious) [inline book/src/proposals/abi-management.md]: Parity

File: book/src/proposals/abi-management.md [hunk 10]
```diff
@@ -0,0 +1,434 @@
...
+  to handle, retrospective in-place update is undesirable at best.
+
+# Solution
+
+Instead of natural human's eye due-diligence, which should be assumed to fail
+regularly, we need a systematic assurance of not breaking the cluster when
+changing the source code.
+
+For that purpose, we propose a two-fold mechanism of marking every ABI-related
+things in source code (`struct`, `enums`) with the new `#[solana_abi]` attribute
```
Comment (garious) [inline book/src/proposals/abi-management.md]: I'd prefer we not prefix anything with `solana`. No need to brand or namespace it. Just call it what it is.

File: book/src/proposals/abi-management.md [hunk 11]
```diff
@@ -0,0 +1,434 @@
...
+and associating ABI-focussed unit tests with them.
+
+The attribute try to detect any unsanctioned changes to the marked ABI-related
```
Comment (garious) [inline book/src/proposals/abi-management.md]: This sentence is tough to read.

File: book/src/proposals/abi-management.md [hunk 12]
```diff
@@ -0,0 +1,434 @@
...
+
+The Solana ABI (binary interface to the cluster) is currently only defined 
+implicitly by the implementation, and requires a very careful eye to notice
+breaking changes. This makes it extremely difficult to upgrade the software
+on an existing cluster without rebooting the ledger.
+
+# Requirements and objectives
+
+- Unintended ABI changes can be detected as CI failures mechanically.
+- Newer impl must be able to process oldest data (since genesis) once we go
```
Comment (mvines) [inline book/src/proposals/abi-management.md]:
```suggestion
- Newer implementation must be able to process oldest data (since genesis) once we go
```

File: book/src/proposals/abi-management.md (hunk 4 above)
Comment (mvines) [reply book/src/proposals/abi-management.md]: I think it holds true.  IMO remove this entire block though: `(This is in ... replayable today)`, it's a little wordy and doesn't really help understanding

File: book/src/proposals/abi-management.md [hunk 13]
```diff
@@ -0,0 +1,434 @@
...
+    processes past object(transactions) exactly in the identical way when
+    replaying the ledger.
+  - Serialization stability: Newer implementation (like AccountsDB) produces
+    newer blobs exactly in the identical way unless its ABI is explicitly and
+    knowingly changed.
+- The objective of this proposal is to protect the ABI while sustaining rather
+  rapid development by opting for mechanical process rather than very long
+  human-driven auditing process. After all, while cluster instability must be
+  avoided, timely delivery of software improvements is desired for all
+  stakeholders of the Solana cluster.
```
Comment (mvines) [inline book/src/proposals/abi-management.md]: > After all, while cluster instability must be avoided, timely delivery of software improvements is desired for all stakeholders of the Solana cluster.

I think this is a given, no need to explicitly call it out :)

File: book/src/proposals/abi-management.md (hunk 2 above)
Comment (mvines) [reply book/src/proposals/abi-management.md]: We may not actually need serialization and deserialization tests.   The solana_abi `digest=` field may be enough.  Once we have the digest computed, it should never change.

File: book/src/proposals/abi-management.md [hunk 14]
```diff
@@ -0,0 +1,434 @@
...
+## Example 2: If we were adding #7415 after some time since mainnet launch:
+
+```patch
+ /// Meta contains enough context to recover the index from storage itself
+ /// This struct will be backed by mmaped and snapshotted data files.
+ /// So the data layout must be stable and consistent across the entire cluster!
+-#[solana_abi(allow=none, digest="662792c9", test="test_hash_stored_account")]
++#[solana_abi(allow=none, digest="c5031113", test="test_hash_stored_account")]
+ #[derive(PartialEq, Debug)]
+ pub struct StoredAccount<'a> {
```
Comment (mvines) [inline book/src/proposals/abi-management.md]: Taking `StoredAccount` as an example here.   I'd prefer that once it's first given a `solana_abi` attribute and digest, it may generally never modified again once the software that first contains its definition ships.   The main exception to this would be if we decide we don't care to retain backwards compatibility for some future release (a deliberate hard fork).

Should a field need to be added or removed, we need to create an entirely new `StoredAccount1` struct (followed by `StoredAccount2`, etc).  At runtime both types will likely be needed.  For example, the old `StoredAccount` is needed when processing older ledger data or dealing with validator running older software.

File: book/src/proposals/abi-management.md (hunk 3 above)
Comment (mvines) [reply book/src/proposals/abi-management.md]: Yes this does feel like overkill to me.

File: book/src/proposals/abi-management.md [hunk 15]
```diff
@@ -0,0 +1,434 @@
...
+  - Implicit version header by optimistic signature verification with version as
+    salt? This detects the version by first verifying signature for blob with
+    (salt=latest revision) and if failed try with (salt=latest revision-1) and
+    so on.
+- how to calculate type digest?
+  - sha1 tuple of (field names, type names), don't consider non derived
+    (De)Serializer; tolerate for incompleteness as mentioned
+- how to integrate well with existing serialization libraries (ie. serde,
+  bincode)?
+- We currently only support amd64, there might be unforseen ABI problems for
```
Comment (mvines) [inline book/src/proposals/abi-management.md]: Yes, great point.   Other architectures too (an ARM port for example).   I'm ok with just saying this is out of scope for the purposes of the current design

File: book/src/proposals/abi-management.md [hunk 16]
```diff
@@ -0,0 +1,434 @@
...
+
+# Considerations
+ 
+- how to detect vesions?
+  - Introduce version header? (2 byte for everything)
+  - Implicit version header by optimistic signature verification with version as
+    salt? This detects the version by first verifying signature for blob with
+    (salt=latest revision) and if failed try with (salt=latest revision-1) and
+    so on.
+- how to calculate type digest?
```
Comment (mvines) [inline book/src/proposals/abi-management.md]: just the types, the field names shouldn't matter for the purposes of an ABI

File: book/src/proposals/abi-management.md (hunk 5 above)
Comment (mvines) [reply book/src/proposals/abi-management.md]: I think version detection can be out of scope for this design.   I see that has a higher-level problem with multiple possible solutions depending on the domain.

For example:
1. Let's say we decide to change the ABI for `Transaction` and now we have `Transaction1`.  To retain backwards compatibility, the validator may continue to accept `Transaction` transactions on the existing TPU port, and create a TPU1 port that accepts `Transaction1` 
2. We create a new inflation model, represented by the new `Inflaction1` struct, that will take affect at Epoch 2020.  Prior to that epoch the bank would use the old `Inflation` struct, and switch over to `Inflation1` in 2020

File: book/src/proposals/abi-management.md (hunk 10 above)
Comment (mvines) [reply book/src/proposals/abi-management.md]: In this case `[solana_abi]` feels pretty accurate.  It's tagging a struct or enum literally is a part of the "Solana ABI"

File: book/src/proposals/abi-management.md (hunk 1 above)
Comment (ryoqun) [reply book/src/proposals/abi-management.md]: With scoping down of this proposal, there is no considerations!

File: book/src/proposals/abi-management.md (hunk 2 above)
Comment (ryoqun) [reply book/src/proposals/abi-management.md]: I see! Thanks for clarification!

File: book/src/proposals/abi-management.md (hunk 3 above)
Comment (ryoqun) [reply book/src/proposals/abi-management.md]: Yeah, I've greatly stripped this down. :)

File: book/src/proposals/abi-management.md (hunk 5 above)
Comment (ryoqun) [reply book/src/proposals/abi-management.md]: Yeah. Thanks for detailed explanation! That's why the original proposal came with some generic concept of `revision` and its associated metadata. Your different use of ports for different ABI has a good point. So, I've changed my mind. :)

No more versioning here!

File: book/src/proposals/abi-management.md (hunk 6 above)
Comment (ryoqun) [reply book/src/proposals/abi-management.md]: As the new proposal shrank considerable, I opted to remove this in its entirely.

File: book/src/proposals/abi-management.md (hunk 7 above)
Comment (ryoqun) [reply book/src/proposals/abi-management.md]: Same as [this](https://github.com/solana-labs/solana/pull/7524#discussion_r360309423).

File: book/src/proposals/abi-management.md (hunk 8 above)
Comment (ryoqun) [reply book/src/proposals/abi-management.md]: This got out-of-scope, too.

File: book/src/proposals/abi-management.md (hunk 9 above)
Comment (ryoqun) [reply book/src/proposals/abi-management.md]: Err, fixed...

File: book/src/proposals/abi-management.md (hunk 11 above)
Comment (ryoqun) [reply book/src/proposals/abi-management.md]: Yeah, I think so too... Now I recomposed [it](https://github.com/solana-labs/solana/pull/7524/files#diff-7bbc1dbd50f4144bb25bfb9d3c28257aR36-R37).

File: book/src/proposals/abi-management.md (hunk 12 above)
Comment (ryoqun) [reply book/src/proposals/abi-management.md]: done

File: book/src/proposals/abi-management.md (hunk 13 above)
Comment (ryoqun) [reply book/src/proposals/abi-management.md]: Removed!

File: book/src/proposals/abi-management.md (hunk 14 above)
Comment (ryoqun) [reply book/src/proposals/abi-management.md]: Thinks for clarification, I've added [a paragraph]https://github.com/solana-labs/solana/pull/7524/files#diff-7bbc1dbd50f4144bb25bfb9d3c28257aR74-R77() accordingly!

File: book/src/proposals/abi-management.md (hunk 15 above)
Comment (ryoqun) [reply book/src/proposals/abi-management.md]: Yeah! I changed to mention this lightly at the start: https://github.com/solana-labs/solana/pull/7524/files#diff-7bbc1dbd50f4144bb25bfb9d3c28257aR42

File: book/src/proposals/abi-management.md (hunk 16 above)
Comment (ryoqun) [reply book/src/proposals/abi-management.md]: Updated!

**Comment by ryoqun** [comment]
> This proposal is quite long. Seems like it'd be sufficient to say, that most modules do not lock down their interfaces to ensure backward compatibility. The few that do use an ad hoc unit-test of a hash of serialized data:
> 
> ```rust
>        //  golden needs to be updated if shred structure changes....
>         let golden: Hash = "2rq8nR6rns2T5zsbQAGBDZb41NVtacneLgkCH17CVxZm"
>             .parse()
>             .unwrap();
> ```
> 
> The drawback of hashing a particular instance of the message is that it doesn't account for any edge cases. Better would be to hash the interface itself. Conveniently `derive(Serialize)` already provides a serialized version of the interface. Proposed solution: add an attribute that accepts a hash, and generate a unit test that verifies the hash of the Serialize data matches the given hash. Add the attribute to all messaging interfaces.

Thanks for reviewing this PR early! Taking your advice, I've significantly narrowed its scope down!

File: book/src/proposals/abi-management.md (hunk 10 above)
Comment (ryoqun) [reply book/src/proposals/abi-management.md]: Naming party here! :)

I'll propose new name `frozen_abi`.

Because....:
- Removing `solana_` makes the name just `abi`, which is too generic.
- Previously I added the `solana_` prefix as a namespace because the whole thing are tightly coupled to the Solana via the revision thing (epochs, cluster version, etc). But this doesn't apply anymore.
- Moreover, I could still use `solana_`, but the new proposal is really small in terms of scope and generic enough to be used by other projects. In fact, it even could be a self-standing generic helper crate around the `serde` (I checked if a similar crate does ever exist or not on the Internet). So no justfication for being under `solana_`'s thing.
- One more thing: `frozen` is trendy these days. :mouse:

**Comment by ryoqun** [comment]
@garious @mvines Phew! I've overhauled the proposal to loss its weight greatly! And I made it out of draft as well! Could you review again?

And thanks for devoting your time for my early work and all your comments are well appreciated!!

**Comment by ryoqun** [comment]
Still, I didn't make this go through Grammarly. So no need to be the grammar person. :)

File: book/src/proposals/abi-management.md [hunk 17]
```diff
@@ -0,0 +1,90 @@
...
+release channel, its digest should never change. If such a change is needed, we
+should opt for defining a new struct like `FooV1`. And special release flow like
+hard forks should be approached.
+
+# Implemention remarks
+
+We use some degree of macro machineries to automatically generate units tests
+and calculate a digest from ABI items. This is doable by clever use of
+`serde::Serialize` ([1]) and `any::typename` ([2]). For a precedent for similar
+implementation, `ink' from the Parity Technologies [3] could be informational.
```
Comment (garious) [inline book/src/proposals/abi-management.md]: That should be a backtick after `ink`

Comment (ryoqun) [reply book/src/proposals/abi-management.md]: Err, fixed!

**Comment by ryoqun** [comment]
> Still, I didn't make this go through Grammarly. So no need to be the grammar person. :)

This is now obsolete! I think much of grammatical errors are fixed by now. :)

**Comment by ryoqun** [comment]
@garious Thanks for checking this out and approvals! Now, I've fixes most of grammarars, I hope. :)

@mvines How does the current form of this PR look from your eyes? :)

**Comment by garious** [comment]
@ryoqun, can you wrap this one up and merge it?

**Comment by ryoqun** [comment]
@garious I see! I merged this! Thanks for reviewing this!
